cardano-crypto-class/src/Cardano/Crypto/DSIGN/Ed25519.hs

```diff
@@ -181,7 +181,9 @@ instance DSIGNAlgorithm Ed25519DSIGN where
     rawSerialiseSigDSIGN      (SigEd25519DSIGN sig) = psbToByteString sig
 
     rawDeserialiseVerKeyDSIGN  = fmap VerKeyEd25519DSIGN . psbFromByteStringCheck
-    rawDeserialiseSignKeyDSIGN = Just . genKeyDSIGN . mkSeedFromBytes
+    rawDeserialiseSignKeyDSIGN bs = do
+      guard (BS.length bs == 32)
```

Comment:
We should do better than just hardcode `32`.
```suggestion
      guard (fromIntegral (BS.length bs) >= seedSizeDSIGN (Proxy @Ed25519DSIGN)))
```
Also it should have at least 32 bytes, not exactly 32. `genKeyDSIGN` documentation says so and when you look at its implementation it actually slices the supplied bytes and only uses the first 32:
https://github.com/input-output-hk/cardano-base/blob/18be4924a2e21261cb1928d8e48384265770f9df/cardano-crypto-class/src/Cardano/Crypto/DSIGN/Ed25519.hs#L162

Reply:
I don't think blindly deserializing more than the expected size and then dropping the rest is good behaviour here.

Reply:
I actually completely agree with you that dropping is not a good behavior, but it was the original design and we need to keep in mind that this is a breaking change. Some code that has worked before might no longer work in the same way. If it was a compilation error I'd be fine, but that will be a different runtime behavior. So we need to investigate what sort of affect it has on downstream users of this function.

It was previously protected for `< 32`, it would just throw a run time error, which is ugly, but it is better than segfault. However it did work fine for `>= 32` bytes, which is a documented feature. So, we can't just change this behavior because we believe it is a good idea, we need to be sure we don't cause any problems while we do that

Reply:
I've doubled checked the cardano-ledger and I believe it is ok to make it more strict and enforce `== 32`, because we never deserialize any signing keys. Although we still do have to change it to use `seedSizeDSIGN`.

@iquerejeta and @michaelpj do you guys have any opinions on this problem? I don't think it should cause any issues for node. But I am just not sure if it is used anywhere else, like in Plutus for example (in particular deserialization of Ed25519 signing keys).

Reply:
Good catch @lehins. Might affect the node as this deserialisation is used in KES, and I wouldn't be surprised if KES key serialisation is > 32 bytes as it not only contains the ed25519 seed, but also the period. It seems that the byte string is split with the correct size, but could someone check?
https://github.com/input-output-hk/cardano-base/blob/1ea947d123981193a6055a193bc2b0e5fd5001ba/cardano-crypto-class/src/Cardano/Crypto/KES/Simple.hs#L177

For Plutus I believe it is fine, as we are still using cardano-crypto there.

Reply:
Yep, that's exactly what I was afraid we could be overlooking. It does affect `rawDeserialiseSignKeyKES` because implementation does chunk up the ByteString into 32 byte long seeds, except the last one, which might be bigger than 32 bytes. We should fix that implementation regardless if we enforce the stricter check or not in `rawDeserialiseSignKeyDSIGN`

Reply:
Actually I take it back `rawDeserialiseSignKeyKES` is not affected, because it will produce exactly `duration` number of 32 byte chunks and if it there is leftover it will produce Nothing. In other words it will fail if the supplied ByteString is not the exact length `sizeVerKeyDSIGN d * t`

Reply:
Glad to see it does not affect KES. But looking at the KES code, it seems that a `SignKeySimpleKES` holds all the remaining secret keys. Am I interpreting that right?
https://github.com/input-output-hk/cardano-base/blob/master/cardano-crypto-class/src/Cardano/Crypto/KES/Simple.hs#L58
@tdammers

Reply:
(not that the above affects these changes, just trying to understand how the KES key is formed)

Reply:
KES key is formed by splitting input bytestring into exactly `duration` number of 32 byte chunks and each individual is converted to `VerKeyDSIGN`, all of which afterwards become `VerKeySimpleKES`. If the supplied ByteString has at least one more or one less byte than expected it will return `Nothing` thanks to this check: `length vkbs == duration`

Reply:
I take that you used the notation of `VerKeySimpleKES`, but the same holds for `SignKeySimpleKES` (so a sign key is converted into exactly `duration` `VerKeyDSIGN` keys).

Reply:
My bad, I meant `SignKeySimpleKES` and `SignKeyDSIGN`. But yeah, `VerKeySimpleKES` has the same logic

Reply:
@lehins , I believe then that we are good to proceed with the strict criteria of size == 32. Or you still want more guarantees that this is not breaking elsewhere?

Reply:
Yeah, I think so. This PR still need this change: `32` -> `seedSizeDSIGN (Proxy @Ed25519DSIGN)`

Other than that if you say its good it is good from my side.